File: frontend/app/create/_lib/agent-events.ts
import type { AgentBundleFile, AgentFileContext, AgentLogItem } from "@/lib/types";

export type FileChange = {
  action: "created" | "modified" | "deleted";
  added: number;
  deleted: number;
  detail?: string;
  diff?: string | null;
  diffFormat?: string;
  line: string;
  path: string;
};

export type AgentContextSummary = {
  files: AgentBundleFile[];
  filesInContext: AgentFileContext[];
  scriptRefs: string[];
};

function eventType(event: unknown): string {
  if (!event || typeof event !== "object") return "";
  const value = (event as { type?: unknown }).type;
  return typeof value === "string" ? value : "";
}

function eventNumber(event: unknown, key: string): number {
  if (!event || typeof event !== "object") return 0;
  const value = (event as Record<string, unknown>)[key];
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

function eventString(event: unknown, key: string): string {
  if (!event || typeof event !== "object") return "";
  const value = (event as Record<string, unknown>)[key];
  return typeof value === "string" ? value : "";
}

function eventArray(event: unknown, key: string): unknown[] {
  if (!event || typeof event !== "object") return [];
  const value = (event as Record<string, unknown>)[key];
  return Array.isArray(value) ? value : [];
}

function eventRecord(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object" && !Array.isArray(value) ? (value as Record<string, unknown>) : null;
}

function optionalEventNumber(record: Record<string, unknown>, key: string): number | undefined {
  const value = record[key];
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

function eventStringArray(value: unknown): string[] {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === "string" && item.length > 0) : [];
}

function bundleFilesFromValue(value: unknown): AgentBundleFile[] {
  if (!Array.isArray(value)) return [];
  return value.flatMap((item): AgentBundleFile[] => {
    const record = eventRecord(item);
    const path = typeof record?.path === "string" ? record.path : "";
    if (!record || !path) return [];
    return [
      {
        bytes: optionalEventNumber(record, "bytes"),
        kind: typeof record.kind === "string" ? record.kind : undefined,
        lines: optionalEventNumber(record, "lines"),
        path,
        referenced: typeof record.referenced === "boolean" ? record.referenced : undefined,
      },
    ];
  });
}

function fileContextsFromValue(value: unknown): AgentFileContext[] {
  if (!Array.isArray(value)) return [];
  return value.flatMap((item): AgentFileContext[] => {
    const record = eventRecord(item);
    const path = typeof record?.path === "string" ? record.path : "";
    if (!record || !path) return [];
    return [
      {
        bytes: optionalEventNumber(record, "bytes"),
        cline_edit_date: optionalEventNumber(record, "cline_edit_date") ?? null,
        cline_read_date: optionalEventNumber(record, "cline_read_date") ?? null,
        deleted: typeof record.deleted === "boolean" ? record.deleted : undefined,
        lines: optionalEventNumber(record, "lines"),
        path,
        record_source: typeof record.record_source === "string" ? record.record_source : undefined,
        record_state: typeof record.record_state === "string" ? record.record_state : undefined,
        updated_at: optionalEventNumber(record, "updated_at"),
      },
    ];
  });
}

export function latestAgentContext(logs: AgentLogItem[]): AgentContextSummary {
  let files: AgentBundleFile[] = [];
  let filesInContext: AgentFileContext[] = [];
  let scriptRefs: string[] = [];

  logs.forEach((log) => {
    logEntries(log).forEach((entry) => {
      const event = eventRecord(entry.event);
      if (!event) return;
      const bundle = eventRecord(event.bundle);
      const bundleFiles = bundleFilesFromValue(bundle?.files);
      const eventFiles = bundleFilesFromValue(event.files);
      const bundleContext = fileContextsFromValue(bundle?.files_in_context);
      const eventContext = fileContextsFromValue(event.files_in_context);
      const bundleRefs = eventStringArray(bundle?.script_refs);
      const eventRefs = eventStringArray(event.script_refs);

      if (bundleFiles.length) files = bundleFiles;
      if (eventFiles.length) files = eventFiles;
      if (bundleContext.length) filesInContext = bundleContext;
      if (eventContext.length) filesInContext = eventContext;
      if (bundleRefs.length) scriptRefs = bundleRefs;
      if (eventRefs.length) scriptRefs = eventRefs;
    });
  });

  return { files, filesInContext, scriptRefs };
}

export function contextSourceLabel(source: string | null | undefined) {
  switch (source) {
    case "read_tool":
      return "read";
    case "cline_edited":
      return "edited";
    case "user_edited":
      return "user edited";
    case "file_mentioned":
      return "mentioned";
    default:
      return source || "tracked";
  }
}

function clineToolLabel(tool: string) {
  switch (tool) {
    case "readFile":
      return "Read file";
    case "searchFiles":
      return "Searched files";
    case "listFilesTopLevel":
    case "listFilesRecursive":
      return "Listed files";
    case "editedExistingFile":
      return "Edited file";
    case "newFileCreated":
      return "Created file";
    case "fileDeleted":
      return "Deleted file";
    case "useSkill":
      return "Read skill";
    default:
      return tool.replace(/_/g, " ");
  }
}

export function activityMessageFromEvent(event: unknown): string {
  const type = eventType(event);
  if (type === "turn_state") {
    const phase = eventString(event, "phase");
    const message = eventString(event, "message");
    const toolCount = eventNumber(event, "tool_count");
    if (phase === "streaming" && toolCount) return `Agent running with ${toolCount} tool(s)`;
    if (phase === "completed") return message || "Agent completed";
    if (phase === "error") return message || "Agent stopped";
    return message || (phase ? `Agent ${phase}` : "");
  }
  if (type === "usage") {
    const total = eventNumber(event, "total_tokens");
    const cached = eventNumber(event, "cached_tokens");
    const cachePercent = eventNumber(event, "cache_percent");
    return total ? `Model usage ${total.toLocaleString()} tokens, cache ${cached.toLocaleString()} (${cachePercent}%)` : "";
  }
  if (type === "file_change") {
    const change = fileChangeFromEvent(event, "");
    if (!change) return "";
    const label = change.action === "created" ? "Created" : change.action === "deleted" ? "Deleted" : "Edited";
    return `${label} ${change.path} (+${change.added} -${change.deleted})`;
  }
  if (type === "heartbeat") {
    const phase = eventString(event, "phase") || "authoring";
    const elapsed = eventNumber(event, "elapsed_seconds");
    const idle = eventNumber(event, "idle_seconds");
    const files = eventNumber(event, "file_count");
    const changed = eventNumber(event, "changed_count");
    const checks = eventString(event, "checks") || "pending";
    return `${phase} waiting on model response: ${elapsed}s elapsed, ${idle}s since last tool, bundle=${files} file(s), changed=${changed}, checks=${checks}`;
  }
  if (type === "check") {
    const checksOk = Boolean((event as Record<string, unknown>).checks_ok);
    const staticErrors = eventNumber(event, "static_errors");
    const smokeOk = Boolean((event as Record<string, unknown>).smoke_ok);
    return `checks ${checksOk ? "passed" : "pending"}: static ${staticErrors ? `${staticErrors} error(s)` : "OK"}, smoke ${smokeOk ? "ok" : "pending"}`;
  }
  if (type === "tool") {
    const clineTool = eventString(event, "cline_tool");
    const tool = clineTool ? clineToolLabel(clineTool) : eventString(event, "tool") || "tool";
    const path = eventString(event, "path") || eventString(event, "name");
    const query = eventString(event, "query");
    const matchCount = eventArray(event, "matches").length;
    const fileCount = eventArray(event, "files").length;
    if (query) return `${tool}: "${query}"${matchCount ? ` (${matchCount} match(es))` : ""}`;
    if (fileCount) return `${tool}: ${fileCount} file(s)`;
    return path ? `${tool} ${path}` : tool;
  }
  if (type === "error") return eventString(event, "message") || "Agent error";
  if (type === "notice") return eventString(event, "message") || "Agent update";
  return "";
}

export function fileChangeFromEvent(event: unknown, line: string): FileChange | null {
  if (eventType(event) !== "file_change") return null;
  const path = eventString(event, "path");
  if (!path) return null;
  const action = eventString(event, "action");
  return {
    action: action === "created" || action === "deleted" ? action : "modified",
    added: eventNumber(event, "added"),
    deleted: eventNumber(event, "deleted"),
    detail: eventString(event, "detail") || undefined,
    diff: eventString(event, "diff") || null,
    diffFormat: eventString(event, "diff_format") || undefined,
    line,
    path,
  };
}

export function logEntries(log: AgentLogItem) {
  if (log.entries?.length) return log.entries;
  const lines = log.lines.length ? log.lines : [log.message];
  return lines.map((line) => ({ line, event: null }));
}

File: backend/app/schemas/agent_events.py
"""Agent 日志事件域:结构化事件联合(discriminator=type)与日志条目。

新事件类型必须同步三处:事件类、AgentLogEventOut 联合、AGENT_LOG_EVENT_TYPES
(serializer 据此丢弃未知类型,防新事件 500 任务端点)。
"""
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field


class AgentBundleFileOut(BaseModel):
    path: str | None = None
    bytes: int | None = None
    lines: int | None = None
    kind: str | None = None
    referenced: bool | None = None


class AgentFileContextOut(BaseModel):
    path: str | None = None
    record_state: str | None = None
    record_source: str | None = None
    bytes: int | None = None
    lines: int | None = None
    deleted: bool | None = None
    updated_at: int | None = None
    cline_read_date: int | None = None
    cline_edit_date: int | None = None


class AgentBundleMetadataOut(BaseModel):
    files: list[AgentBundleFileOut] | None = None
    script_refs: list[str] | None = None
    files_in_context: list[AgentFileContextOut] | None = None


class AgentSearchMatchOut(BaseModel):
    path: str | None = None
    line: int | None = None
    text: str | None = None


class _AgentLogEventBaseOut(BaseModel):
    model_config = ConfigDict(extra="allow")

    seq: int | None = None
    status: str | None = None


class AgentFileChangeEventOut(_AgentLogEventBaseOut):
    type: Literal["file_change"]
    action: Literal["created", "modified", "deleted"] | None = None
    path: str | None = None
    added: int | None = None
    deleted: int | None = None
    bytes: int | None = None
    chunks: int | None = None
    detail: str | None = None
    diff: str | None = None
    diff_format: Literal["unified", "omitted_large", "empty"] | None = None
    cline_tool: str | None = None
    files_in_context: list[AgentFileContextOut] | None = None
    tool: str | None = None


class AgentTurnStateEventOut(_AgentLogEventBaseOut):
    type: Literal["turn_state"]
    phase: str | None = None
    message: str | None = None
    source: str | None = None
    reason: str | None = None
    agent: str | None = None
    tool_count: int | None = None
    bundle: AgentBundleMetadataOut | None = None
    checks_ok: bool | None = None
    changed: list[str] | None = None


class AgentHeartbeatEventOut(_AgentLogEventBaseOut):
    type: Literal["heartbeat"]
    agent: str | None = None
    operation: Literal["authoring", "repairing"] | None = None
    phase: str | None = None
    elapsed_seconds: int | None = None
    idle_seconds: int | None = None
    file_count: int | None = None
    changed_count: int | None = None
    checks: str | None = None
    files_in_context: list[AgentFileContextOut] | None = None


class AgentCheckEventOut(_AgentLogEventBaseOut):
    type: Literal["check"]
    tool: str | None = None
    static_ok: bool | None = None
    static_errors: int | None = None
    smoke_ok: bool | None = None
    checks_ok: bool | None = None
    bundle: AgentBundleMetadataOut | None = None


class AgentToolEventOut(_AgentLogEventBaseOut):
    type: Literal["tool"]
    tool: str | None = None
    cline_tool: str | None = None
    path: str | None = None
    name: str | None = None
    bytes: int | None = None
    query: str | None = None
    file_pattern: str | None = None
    files: list[AgentBundleFileOut] | None = None
    script_refs: list[str] | None = None
    files_in_context: list[AgentFileContextOut] | None = None
    matches: list[AgentSearchMatchOut] | None = None


class AgentUsageEventOut(_AgentLogEventBaseOut):
    type: Literal["usage"]
    input_tokens: int | None = None
    output_tokens: int | None = None
    total_tokens: int | None = None
    cached_tokens: int | None = None
    cache_write_tokens: int | None = None
    requests: int | None = None
    cache_percent: int | None = None


class AgentUsageProgressEventOut(_AgentLogEventBaseOut):
    type: Literal["usage_progress"]
    agent: str | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    total_tokens: int | None = None
    cached_tokens: int | None = None
    cache_write_tokens: int | None = None


class AgentErrorEventOut(_AgentLogEventBaseOut):
    type: Literal["error"]
    message: str | None = None
    source: str | None = None


class AgentNoticeEventOut(_AgentLogEventBaseOut):
    type: Literal["notice"]
    message: str | None = None
    reason: str | None = None


class AgentRoleBudgetExhaustedEventOut(_AgentLogEventBaseOut):
    type: Literal["role_budget_exhausted"]
    agent: str
    operation: Literal["authoring", "repairing"]
    reason: Literal["max_turns"] = "max_turns"
    message: str | None = None
    turns_limit: int | None = None
    changed: list[str] | None = None
    checks_ok: bool | None = None


class AgentRepairAttemptStartedEventOut(_AgentLogEventBaseOut):
    type: Literal["repair_attempt_started"]
    agent: str | None = None
    operation: Literal["repairing"] = "repairing"
    repair_kind: Literal["build", "revision", "gameplay"] | None = None
    attempt: int | None = None
    max_attempts: int | None = None
    caused_by_step_id: str | None = None


class AgentAuthorTeamEventOut(_AgentLogEventBaseOut):
    type: Literal["author_team"]
    phase: str | None = None
    role: str | None = None
    message: str | None = None
    base_revision: str | None = None
    contract_hash: str | None = None
    contract_source: str | None = None


class AgentPatchDiagnosticOut(BaseModel):
    model_config = ConfigDict(extra="allow")

    code: str | None = None
    path: str | None = None
    line: int | None = None
    column: int | None = None
    rule: str | None = None
    message: str | None = None


class AgentValidationRejectionEventOut(_AgentLogEventBaseOut):
    type: Literal["validation_rejection"]
    tool: str | None = None
    path: str | None = None
    diagnostics: list[AgentPatchDiagnosticOut] | None = None
    diagnostics_omitted: int | None = None


class AgentRetryEventOut(_AgentLogEventBaseOut):
    type: Literal["retry"]
    source: str | None = None
    attempt: int | None = None
    next_attempt: int | None = None
    reason: str | None = None
    stream_event: str | None = None
    error_code: str | None = None
    response_id: str | None = None
    delay_seconds: float | None = None


class AgentRoleStreamFailedPartialEventOut(_AgentLogEventBaseOut):
    type: Literal["role_stream_failed_partial"]
    agent: str | None = None
    operation: Literal["authoring", "repairing"] | None = None
    reason: str | None = None
    message: str | None = None
    changed: list[str] | None = None
    checks_ok: bool | None = None


AgentLogEventOut = Annotated[
    AgentFileChangeEventOut
    | AgentTurnStateEventOut
    | AgentHeartbeatEventOut
    | AgentCheckEventOut
    | AgentToolEventOut
    | AgentUsageEventOut
    | AgentUsageProgressEventOut
    | AgentErrorEventOut
    | AgentNoticeEventOut
    | AgentRoleBudgetExhaustedEventOut
    | AgentRepairAttemptStartedEventOut
    | AgentAuthorTeamEventOut
    | AgentValidationRejectionEventOut
    | AgentRetryEventOut
    | AgentRoleStreamFailedPartialEventOut,
    Field(discriminator="type"),
]

# Single source of truth for serializers: payloads whose type is not listed
# here must be dropped before response validation, never allowed to 500 the
# task endpoints when an agent starts emitting a new event type.
AGENT_LOG_EVENT_TYPES = frozenset(
    {
        "file_change",
        "turn_state",
        "heartbeat",
        "check",
        "tool",
        "usage",
        "usage_progress",
        "error",
        "notice",
        "role_budget_exhausted",
        "repair_attempt_started",
        "author_team",
        "validation_rejection",
        "retry",
        "role_stream_failed_partial",
    }
)


class AgentLogEntryOut(BaseModel):
    cursor: int | None = None
    line: str
    level: str | None = None
    created_at: str | None = None
    event: AgentLogEventOut | None = None


class AgentLogItemOut(BaseModel):
    step_id: str | None = None
    agent_name: str
    step: str
    message: str
    created_at: str | None = None
    duration: str | None = None
    status: str
    lines: list[str]
    entries: list[AgentLogEntryOut] | None = None
